File: Problems/Algorithms/438. Find All Anagrams in a String/FindAnagrams.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] counts = new int[128];
        for (char ch : p.toCharArray()) counts[ch]++;
        
        List<Integer> results = new ArrayList<>();
        for (int right = 0, left = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            counts[ch]--;
            while (counts[ch] < 0) {
                counts[s.charAt(left)]++;
                left++;
            }

            if (right - left + 1 == p.length()) {
                results.add(left);
            }
        }
        return results;
    }
}

File: Problems/Algorithms/438. Find All Anagrams in a String/FindAnagrams.kt
/* The following solutions incurs TLE */
class Solution {
    fun findAnagrams(s: String, p: String): List<Int> {
        val map: MutableMap<String, MutableList<Int>> = mutableMapOf()
        
        val n = s.length
        val m = p.length
        if (n < m) return listOf()
        
        for (i in 0..n-m) {
            val curr = s.substring(i, i+m).toCharArray().sorted().joinToString("")
            if (!map.containsKey(curr)) {
                map.put(curr, mutableListOf(i))
            } else {
                map.get(curr)!!.add(i)
            }
        }
        
        return map.get(p.toCharArray().sorted().joinToString("")) ?: listOf()
    }
}

